<a href="https://colab.research.google.com/github/shwetakathait/practice_dataset/blob/main/XGBoost_Covid_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# 1.0 Call libraries:

import pandas as pd
from  pathlib import Path

# 1.0.1
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

# 1.0.2
from xgboost import XGBRegressor
from sklearn.metrics import accuracy_score, f1_score

In [ ]:
# 2.0 Read the data
data = pd.read_excel(r"C:\Users\verma\Downloads\covid_data.xlsx")

In [ ]:
# 2.1 Show data:

data.shape
data.head()

Continent                Country  Population  CasesNew  CasesActive  \
0  North-America               Anguilla     15230.0       NaN         13.0   
1  North-America  Saint-Pierre-Miquelon      5759.0       NaN       1001.0   
2         Africa           Saint-Helena      6115.0       NaN       2164.0   
3  South-America       Falkland-Islands      3539.0       NaN          0.0   
4  North-America             Montserrat      4965.0       NaN         19.0   

   CasesCritical  CasesRecovered  Cases.1M_pop  CasesTotal  DeathsNew  \
0            4.0          3879.0      256336.0        3904        NaN   
1            1.0          2449.0      599410.0        3452        NaN   
2            NaN             2.0      354211.0        2166        NaN   
3            NaN          1930.0      545352.0        1930        NaN   
4            NaN          1376.0      282578.0        1403        NaN   

   Deaths.1M_pop  DeathsTotal  Tests.1M_pop  TestsTotal        Day  \
0          788.0         12.0     3373736.0     51382.0 2023-07-07   
1          347.0          2.0     4410488.0     25400.0 2023-07-07   
2            NaN          NaN           NaN         NaN 2023-07-07   
3            NaN          NaN     2439107.0      8632.0 2023-07-07   
4         1611.0          8.0     3577442.0     17762.0 2023-07-07   

                        Time  
0  2023-07-07T17:30:09+00:00  
1  2023-07-07T17:30:09+00:00  
2  2023-07-07T17:30:09+00:00  
3  2023-07-07T17:30:09+00:00  
4  2023-07-07T17:30:09+00:00

In [ ]:
df = data.dropna()

In [ ]:
# 2.2 Select subset of predictors:

cols_to_use = ['Population', 'CasesNew', 'CasesActive', 'CasesRecovered']
X = df[cols_to_use]

# 2.3 Select target
y = df.DeathsTotal

# 2.4 Separate data into training and validation sets
X_train, X_valid, y_train, y_valid = train_test_split(X, y)

In [ ]:
# 3.0
my_model = XGBRegressor()
my_model.fit(X_train, y_train, verbose = True)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

In [ ]:
# 3.1 Predictions:

predictions = my_model.predict(X_valid)
print("Mean Absolute Error: " + str(mean_absolute_error(predictions, y_valid)))

Mean Absolute Error: 12685.271606445312


In [ ]:
# 3.2 Set n_estimators:

my_model = XGBRegressor(n_estimators=500)
my_model.fit(X_train, y_train, verbose = True)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=500, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

In [ ]:
# 3.3 Set early stop rounds:

my_model = XGBRegressor(n_estimators=500)
my_model.fit(X_train, y_train,
             early_stopping_rounds=5,
             eval_set=[(X_valid, y_valid)],
             verbose= True)

[0]	validation_0-rmse:66419.75102
[1]	validation_0-rmse:58541.06439
[2]	validation_0-rmse:51810.37505
[3]	validation_0-rmse:46117.47710
[4]	validation_0-rmse:41316.87081
[5]	validation_0-rmse:37295.23971
[6]	validation_0-rmse:33862.96422
[7]	validation_0-rmse:30966.87038
[8]	validation_0-rmse:28524.98060
[9]	validation_0-rmse:26548.78884
[10]	validation_0-rmse:24889.39691
[11]	validation_0-rmse:23496.56879
[12]	validation_0-rmse:22328.09027
[13]	validation_0-rmse:21347.93404
[14]	validation_0-rmse:20520.45975
[15]	validation_0-rmse:19824.90002
[16]	validation_0-rmse:19240.08848
[17]	validation_0-rmse:18747.98943
[18]	validation_0-rmse:18333.60031
[19]	validation_0-rmse:17984.38166
[20]	validation_0-rmse:17689.85506
[21]	validation_0-rmse:17441.25698
[22]	validation_0-rmse:17231.26982
[23]	validation_0-rmse:17053.76991
[24]	validation_0-rmse:16903.27422
[25]	validation_0-rmse:16775.89639
[26]	validation_0-rmse:16668.02268
[27]	validation_0-rmse:16576.62426
[28]	validation_0-rmse:16499.2

C:\Users\verma\anaconda3\lib\site-packages\xgboost\sklearn.py:835: UserWarning: `early_stopping_rounds` in `fit` method is deprecated for better compatibility with scikit-learn, use `early_stopping_rounds` in constructor or`set_params` instead.
  warnings.warn(


XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=500, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

In [ ]:
# 3.4 learning_rate:

my_model = XGBRegressor(n_estimators=1000, learning_rate=0.05)
my_model.fit(X_train, y_train,
             early_stopping_rounds=5,
             eval_set=[(X_valid, y_valid)],
             verbose= True)

[0]	validation_0-rmse:74219.29125
[1]	validation_0-rmse:72697.90709
[2]	validation_0-rmse:71216.66288
[3]	validation_0-rmse:69774.49070
[4]	validation_0-rmse:68370.36034
[5]	validation_0-rmse:67003.26339
[6]	validation_0-rmse:65672.21599
[7]	validation_0-rmse:64376.26918
[8]	validation_0-rmse:63106.87868
[9]	validation_0-rmse:61864.89602
[10]	validation_0-rmse:60655.51831
[11]	validation_0-rmse:59483.36552
[12]	validation_0-rmse:58336.42118
[13]	validation_0-rmse:57219.62488
[14]	validation_0-rmse:56137.38005
[15]	validation_0-rmse:55078.32222
[16]	validation_0-rmse:54047.14912
[17]	validation_0-rmse:53043.14795
[18]	validation_0-rmse:52070.27766
[19]	validation_0-rmse:51115.11063
[20]	validation_0-rmse:50184.55869
[21]	validation_0-rmse:49278.00095
[22]	validation_0-rmse:48394.81818
[23]	validation_0-rmse:47534.42750
[24]	validation_0-rmse:46696.24675
[25]	validation_0-rmse:45879.70532
[26]	validation_0-rmse:45084.25946
[27]	validation_0-rmse:44309.37013
[28]	validation_0-rmse:43557.4

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.05, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=1000, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

In [ ]:
# n_jobs:

my_model = XGBRegressor(n_estimators=1000, learning_rate=0.05, n_jobs=4)
my_model.fit(X_train, y_train,
             early_stopping_rounds=5,
             eval_set=[(X_valid, y_valid)],
             verbose= True)

[0]	validation_0-rmse:74219.29125
[1]	validation_0-rmse:72697.90709
[2]	validation_0-rmse:71216.66288
[3]	validation_0-rmse:69774.49070
[4]	validation_0-rmse:68370.36034
[5]	validation_0-rmse:67003.26339
[6]	validation_0-rmse:65672.21599
[7]	validation_0-rmse:64376.26918
[8]	validation_0-rmse:63106.87868
[9]	validation_0-rmse:61864.89602
[10]	validation_0-rmse:60655.51831
[11]	validation_0-rmse:59483.36552
[12]	validation_0-rmse:58336.42118
[13]	validation_0-rmse:57219.62488
[14]	validation_0-rmse:56137.38005
[15]	validation_0-rmse:55078.32222
[16]	validation_0-rmse:54047.14912
[17]	validation_0-rmse:53043.14795
[18]	validation_0-rmse:52070.27766
[19]	validation_0-rmse:51115.11063
[20]	validation_0-rmse:50184.55869
[21]	validation_0-rmse:49278.00095
[22]	validation_0-rmse:48394.81818
[23]	validation_0-rmse:47534.42750
[24]	validation_0-rmse:46696.24675
[25]	validation_0-rmse:45879.70532
[26]	validation_0-rmse:45084.25946
[27]	validation_0-rmse:44309.37013
[28]	validation_0-rmse:43557.4

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.05, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=1000, n_jobs=4, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

Analysis and Interpretation

# Objective
Predicting the number of deaths based on different predictors (features) including population, new cases, active cases, and recovered cases by performing regression analysis with the XGBoost algorithm.

# Brief
•	The code shows how to train an XGBoost regression model in a variety of ways, including defining hyperparameters, early halting, modifying learning rates, and using multiple cores. The main goal is to create a model that correctly forecasts the number of deaths using the information provided as predictor variables. The model's performance is evaluated using evaluation measures like MAE and early halting.

•	Selecting predictors and target: We took a subset of predictors ('Population', 'CasesNew', 'CasesActive', 'CasesRecovered') and assigned them to the variable X.
The target variable ('DeathsTotal') is assigned to the variable y.

•	Early stopping: Early stopping is implemented to prevent overfitting. The early stopping_rounds parameter specifies that if the model's performance on the validation set (X_valid, y_valid) doesn't improve for 5 consecutive rounds, the training will stop.

# Outcome
The model's predictions are made on the validation data (X_valid) and compared to the actual target values (y_valid). The mean absolute error (MAE) is calculated and printed, which provides an indication of how well the model's predictions match the actual values. The MAE value, 12685.271606445312 represents the average absolute difference between the predicted number of deaths and the actual number of deaths across the validation set.
